test/e2e/upgrades/mysql.go

```diff
@@ -0,0 +1,257 @@
+		framework.Logf("Attempt number %d at finding write Service IP.", numRetries+1)
+		service, err := f.ClientSet.CoreV1().Services(ns).Get("mysql-write", metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		ingress := service.Status.LoadBalancer.Ingress
+		if len(ingress) > 0 {
+			ip = ingress[0].IP
+			continue
+		}
+		time.Sleep(time.Second * 10)
+	}
+
+	By("Connecting to the database using the write Service.")
+	db, err := sql.Open("mysql", "root@tcp("+ip+":3306)/")
+	Expect(err).NotTo(HaveOccurred())
+
+	err = db.Ping()
+	Expect(err).NotTo(HaveOccurred())
+
+	By("Inserting some basic data into the database.")
+
+	res, err := db.Exec("CREATE DATABASE testing;")
+	Expect(err).NotTo(HaveOccurred())
+	framework.Logf("res was: %v", res)
+
+	res, err = db.Exec("CREATE TABLE testing.users (name VARCHAR(150) NOT NULL, PRIMARY KEY (name));")
+	Expect(err).NotTo(HaveOccurred())
+	framework.Logf("res was: %v", res)
+
+	res, err = db.Exec("INSERT INTO testing.users (name) VALUES(?);", "Alice")
+	Expect(err).NotTo(HaveOccurred())
+	framework.Logf("res was: %v", res)
+
+	res, err = db.Exec("INSERT INTO testing.users (name) VALUES(?);", "Bob")
+	Expect(err).NotTo(HaveOccurred())
+	framework.Logf("res was: %v", res)
+
+	rows, err := db.Query("SELECT * FROM testing.users;")
+	Expect(err).NotTo(HaveOccurred())
+	framework.Logf("rows was: %v", rows)
+
+	db.Close()
+
+	By("Waiting for the read Service's external IP to be available.")
+	for numRetries := 0; numRetries < maxRetries; numRetries++ {
+		framework.Logf("Attempt number %d at finding read Service IP.", numRetries+1)
+		service, err := f.ClientSet.CoreV1().Services(ns).Get("mysql-read", metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		ingress := service.Status.LoadBalancer.Ingress
+		if len(ingress) > 0 {
+			ip = ingress[0].IP
+			continue
+		}
+		time.Sleep(time.Second * 10)
+	}
+
+	By("Connecting to the database using the read service.")
+	db, err = sql.Open("mysql", "root@tcp("+ip+":3306)/")
+	Expect(err).NotTo(HaveOccurred())
+	t.db = db
+
+	err = t.db.Ping()
+	Expect(err).NotTo(HaveOccurred())
+
+	By("Confirming the data is available before starting the upgrade.")
+	err = t.CheckRows()
+	Expect(err).NotTo(HaveOccurred())
+}
+
+// Test continually polls the db using the read Service connection, checking to see if the data
+// inserted earlier is available.
+func (t *MySqlUpgradeTest) Test(f *framework.Framework, done <-chan struct{}, upgrade UpgradeType) {
+	By("Continuously polling the database during upgrade.")
+	wait.Until(func() {
+		err := t.CheckRows()
+		if err != nil {
+			framework.Logf("Error while trying to confirm data: %v", err)
+			t.failure = t.failure + 1
+		} else {
+			t.success = t.success + 1
+		}
+	}, framework.Poll, done)
+
+	// TODO: What should success be? What number of failures? The only failures i've
+	// noticed so far are from race conditions (lookup for data starts, doesn't finish before
+	// upgrade interferes).
```

Comment:
That's fine as a starting point.  I would pick a value in that neighborhood and leave the TODO comment.  We should look at what real life failure/success rates are after it's run for awhile.